gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/base/numericalDerivative.h>
+
+
+namespace gtsam {
+
+    /**
+     * @class PlanarProjectionFactorBase
+     * @brief Camera projection for robot on the floor.  Measurement is camera pixels.
+     */
+    class PlanarProjectionFactorBase {
+    protected:
+        PlanarProjectionFactorBase() {}
+
+        /**
+         * @param measured corresponding point in the camera frame
+         * @param poseKey index of the robot pose in the z=0 plane
+         */
+        PlanarProjectionFactorBase(const Point2& measured) : measured_(measured) {}
+
+        /**
+         * @param landmark point3
+         * @param pose
+         * @param offset oriented parallel to pose2 zero i.e. +x
+         * @param calib
+         * @param H1 landmark jacobian
+         * @param H2 pose jacobian
+         * @param H3 offset jacobian
+         * @param H4 calib jacobian
+         */
+        Point2 h(
+            const Point3& landmark,
+            const Pose2& pose,
+            const Pose3& offset,
+            const Cal3DS2& calib,
+            OptionalMatrixType H1, // 2x3 (x, y, z)
+            OptionalMatrixType H2, // 2x3 (x, y, theta)
+            OptionalMatrixType H3, // 2x6 (rx, ry, rz, x, y, theta)
+            OptionalMatrixType H4  // 2x9
+        ) const {
+            try {
+                // this is x-forward z-up
+                gtsam::Matrix H0; // 6x6
+                Pose3 offset_pose = Pose3(pose).compose(offset, H0);
+                // this is z-forward y-down
+                gtsam::Matrix H00; // 6x6
+                Pose3 camera_pose = offset_pose.compose(CAM_COORD, H00);
+                PinholeCamera<Cal3DS2> camera = PinholeCamera<Cal3DS2>(camera_pose, calib);
+                if (H2 || H3) {
+                    // Dpose is for pose3, 2x6 (R,t)
+                    gtsam::Matrix Dpose;
+                    Point2 result = camera.project(landmark, Dpose, H1, H4);
+                    gtsam::Matrix DposeOffset = Dpose * H00; // 2x6
+                    if (H3)
+                        *H3 = DposeOffset; // with Eigen this is a deep copy (!)
+                    if (H2) {
+                        gtsam::Matrix DposeOffsetFwd = DposeOffset * H0;
+                        *H2 = Matrix::Zero(2, 3);
+                        (*H2)(0, 0) = DposeOffsetFwd(0, 3); // du/dx
+                        (*H2)(1, 0) = DposeOffsetFwd(1, 3); // dv/dx
+                        (*H2)(0, 1) = DposeOffsetFwd(0, 4); // du/dy
+                        (*H2)(1, 1) = DposeOffsetFwd(1, 4); // dv/dy
+                        (*H2)(0, 2) = DposeOffsetFwd(0, 2); // du/dyaw
+                        (*H2)(1, 2) = DposeOffsetFwd(1, 2); // dv/dyaw
+                    }
+                    return result;
+                } else {
+                    return camera.project(landmark, {}, {}, {});
+                }
+            } catch (CheiralityException& e) {
```

Comment:
We have a typedef `GTSAM_THROW_CHEIRALITY_EXCEPTION`. Please ifdef your catch based on that.

Reply:
done